### PR TITLE
Upgrade ESLint and dependencies, fix new lint errors, switch Travis to Yarn

### PR DESCRIPTION
**Pull request by tomocchino**
Test Plan: yarn run lint, yarn run test, flow, everything is clean

This also involved pulling in the latest fbjs so we can extend that .eslintrc.js. One goal here is to be able to use yarn exclusively in Travis / CircleCI.

As a followup, I can try to add a custom rule to turn the ESLint `strict` rule back on, because we need to deal with the fact that ESLint thinks `import type` Flow statements are module imports.

Sorry this is doing so much, but these are all pretty interconnected. Let me know if you want me to split this into multiple PRs somehow.

### CODE REVIEW

File: src/renderers/testing/ReactTestRenderer.js [hunk 1]
```diff
@@ -97,6 +97,8 @@ class ReactTestComponent {
   }
 
   toJSON(): ReactTestRendererJSON {
+    // not using `children`, but I don't wanna rewrite without destructuring
```
Comment (zpao) [inline src/renderers/testing/ReactTestRenderer.js]: wanna?

**Comment by gaearon** [comment]
Tagging as patch because it also happens to fix an issue for Rollup users. (https://github.com/facebook/react/issues/8318)
The bug is Rollup's but it's easy for us to fix by shipping this commit.

**Comment by zpao** [comment]
I haven't tried but this might be pretty tough to cherry-pick in a patch, perhaps even a minor. It might be worthwhile to just take the specific changes in the branch.

**Comment by gaearon** [comment]
k, I'll cherry-pick it

**Pull request merged**